rkt run: rework CLI app/image parsing

vcaputo opened the pull request:
This builds on top of #668 moving Apps into a pkg with an eye towards using the Apps struct to pass the cli-derived state into stage0 rather than the various lists.

eyakubovich commented on common/apps/apps.go:
```diff
@@ -0,0 +1,140 @@
+	imageID types.Hash // resolved image identifier
+}
+
+type Apps struct {
```
I think it's more Go idiomatic to make this a list subtype: `type Apps []App`

vcaputo replied:
More stuff goes into the struct later, basically the global flags will populate members of the Apps struct.  Not to mention the rules for method declarations got in the way of making it work the way you're suggesting.  The only way I could get it to compile had the flag-specific Set() function operating on its own slice of Apps, not a pointer to the one we need to expand and access later...

jonboulle replied:
I really don't understand, can you give an example please?

vcaputo replied:
@jonboulle I was incorrect, after playing around with putting together a simpler test I found what I was doing incorrectly, it's totally doable.

Regardless the struct is useful for putting the global flags into, the volumes for example can go there, the global mounts, the global set-envs, all that stuff will land in the Apps struct.

jonboulle replied:
Ah! Gotcha, I misunderstood what you meant by "more stuff goes into [it] later". SGTM

eyakubovich commented on common/apps/apps.go:
```diff
@@ -0,0 +1,140 @@
+	asc   string   // signature file override for image verification (if fetching occurs)
+
+	imageID types.Hash // resolved image identifier
+}
```
Since these are exposed via Get/Set, I would just make them public and not do the Javaisque getters and setters.

jonboulle replied:
+1

eyakubovich commented on rkt/fetch.go:
```diff
@@ -111,45 +143,53 @@ func fetchImage(img string, ds *cas.Store, ks *keystore.Keystore, discover bool)
-	case "http", "https", "docker":
+	case "http", "https", "docker", "file":
 	default:
 		return "", fmt.Errorf("rkt only supports http, https or docker URLs (%s)", img)
```
add file URLs

vcaputo replied:
I don't think it's necessary to expose to the user that file URLs are supported, we internally generate file urls when an image path which exists in the filesystem is provided, it's a bit of an implementation detail.  Does it make sense to add more verbosity to the message here?

jonboulle replied:
It makes sense to at least mention that file references are valid..

jonboulle replied:
https://github.com/coreos/rocket/issues/710

eyakubovich commented:
LGTM other than nits above

vcaputo commented:
@jon

jonboulle commented on common/apps/apps.go:
```diff
@@ -0,0 +1,140 @@
+)
+
+type App struct {
+	image string   // the image as-supplied by the user on the cli
```
s#as-supplied#as supplied#
maybe "image reference"?

jonboulle commented on common/apps/apps.go:
```diff
@@ -0,0 +1,140 @@
+func (al *Apps) Last() *App {
+	if len(al.apps) == 0 {
+		return nil
+	} else {
```
prefer shortcircuit to if/else

jonboulle commented on common/apps/apps.go:
```diff
@@ -0,0 +1,140 @@
+	for i, _ := range al.apps {
+		// FIXME(vc): is there a way to make:
+		//  for _, a := range al.apps
+		// produce a pointer to the app in 'a'?
```
For better or worse, the idiomatic approach is
```
for _, a := range al.apps {
   a := a
   if err := f(&a); err != nil {
```

jonboulle commented on common/apps/apps.go:
```diff
@@ -0,0 +1,140 @@
+// these convenience functions just return typed lists containing just the named member
+// TODO(vc): these probably go away when we just pass Apps to stage0
+
+// getImages returns a list of the images in al, one per app.
```
nit, capitalisation of docstrings here and below is inconsistent with actual function names

jonboulle commented on rkt/cli_apps.go:
```diff
@@ -0,0 +1,126 @@
+// parseApps looks through the args for support of per-app argument lists delimited with "--" and "---".
+// Between per-app argument lists flags.Parse() is called using the supplied FlagSet.
+// Anything not consumed by flags.Parse() and not found to be a per-app argument list is treated as an image.
+func parseApps(al *apps.Apps, args []string, flags *flag.FlagSet) error {
```
would love to see what the coverage looks like for this function :-)

jonboulle commented on rkt/cli_apps.go:
```diff
@@ -0,0 +1,126 @@
+	}
+
+	app.SetAsc(s)
+
```
kill some of the superfluous newlines

jonboulle commented on stage0/run.go:
```diff
@@ -51,7 +52,7 @@ import (
 type PrepareConfig struct {
 	CommonConfig
 	// TODO(jonboulle): These images are partially-populated hashes, this should be clarified.
```
uhh, somewhere in a reshuffle this TODO lost its home, can you find it again?

jonboulle replied:
ah, line 68

vcaputo replied:
it's back home again

jonboulle commented on rkt/rkt.go:
```diff
@@ -53,11 +53,11 @@ func init() {
+	Summary     string        // One-sentence summary of what the Command does
+	Usage       string        // Usage options/arguments
+	Description string        // Detailed description of command
+	Flags       *flag.FlagSet // Set of flags associated with this command
```
Can you explain the necessity of this change please?

vcaputo replied:
The circular initialization upsets Go, we use Command.Flags from the run$Cmd function via parseApps() and Command.Run is initialized with run$Cmd.

If you know a better way to work around it, I'm not happy with this, just working around the compiler.

jonboulle replied:
ah, missed that. I remember tackling this aaaaaaages ago with babby's first CLI library, can't remember if I ever came up with a decent solution though, will ponder. (maybe `Run func(fs *flags.FlagSet, args []string) int` or something)

jonboulle commented on rkt/rkt.go:
```diff
@@ -100,6 +100,7 @@ func main() {
 		log.SetOutput(os.Stderr)
 	}
 
+	// FIXME(vc): Flags.Args() swallows the "--" which may have stopped its parsing, this interferes with parseApps()
```
Should this be fixed before merging?

vcaputo replied:
Probably, it just prevents a graceful exit/error message when someone does something like:
`rkt run -- args for image` where we should error because there's no image yet, but `rkt run` subcommand just gets `args for image`.

I can do a similar check to what I do in parseApps() to see if `--` halted the parsing, and put it back so the subcommand receives it.

jonboulle commented on rkt/run.go:
```diff
@@ -50,8 +43,17 @@ They will be checked in that order and the first match will be used.
 An "--" may be used to inhibit rkt run's parsing of subsequent arguments,
 which will instead be appended to the preceding image app's exec arguments.
 End the image arguments with a lone "---" to resume argument parsing.`,
```
I would love to see some full examples in this help output

jonboulle commented on rkt/run.go:
```diff
@@ -206,25 +109,19 @@ func runRun(args []string) (exit int) {
-	ks := getKeystore()
 
-	s1img, err := findImage(flagStage1Image, ds, ks, false)
+	s1img, err := findImage(flagStage1Image, "", ds, nil, false)
```
does it make sense to support --signature for the stage1?!?

vcaputo replied:
It's not a regression, we weren't verifying stage1 at all previously because local files didn't go through verification.

Let's address that in a separate PR, it will probably get more complicated because I suspect we'll want to pin the default stage1 to a specific imageid/sha so unless your running with a stage1 flag it doesn't need to verify the stage1 on every startup without `--insecure-skip-verify`.

jonboulle replied:
Yep I'm fine with that, was just raising it as an idea

jonboulle commented on rkt/run.go:
```diff
@@ -206,25 +109,19 @@ func runRun(args []string) (exit int) {
 
-	imgs, err := findImages(images, ds, ks)
-	if err != nil {
+	ks := getKeystore()
```
can just inline this in 120

jonboulle commented on rkt/prepare.go:
```diff
@@ -61,13 +64,12 @@ func runPrepare(args []string) (exit int) {
-	appArgs, images, err := parseAppArgs(args)
-	if err != nil {
-		stderr("prepare: error parsing app image arguments")
+	if err = parseApps(&rktApps, args, &prepareFlags); err != nil {
```
Any particular reason for 
```
func parseApps(al *apps.Apps, args []string, flags *flag.FlagSet) error
```
vs.
```
func parseApps(args []string, flags *flag.FlagSet) (*apps.Apps, error) 
```
?

vcaputo replied:
tests and the ability to potentially seed global stuff in *apps.Apps before calling parseApps()

jonboulle commented on rkt/images.go:
```diff
@@ -0,0 +1,76 @@
+
+// findImages uses findImage to attain a list of image hashes using discovery if necessary
+func findImages(al *apps.Apps, ds *cas.Store, ks *keystore.Keystore) error {
+	if err := al.Walk(func(app *apps.App) error {
```
For cases like this, please assign then check in a new conditional
```
err := al.Walk ...
if err != nil
```

... oh actually or you can just return err directly

jonboulle commented on rkt/fetch.go:
```diff
@@ -71,7 +74,7 @@ func runFetch(args []string) (exit int) {
 
 	for _, img := range args {
-		hash, err := fetchImage(img, ds, ks, true)
+		hash, err := fetchImage(img, "" /* TODO(vc): wire to --signature */, ds, ks, true)
```
todo when? :-)

jonboulle commented on rkt/fetch.go:
```diff
@@ -85,9 +88,38 @@ func runFetch(args []string) (exit int) {
 // fetchImage will take an image as either a URL or a name string and import it
 // into the store if found.  If discover is true meta-discovery is enabled.
-func fetchImage(img string, ds *cas.Store, ks *keystore.Keystore, discover bool) (string, error) {
+// if asc is not "", it must exist as a local file and will be used as the signature file for verification unless verification is disabled.
```
s#if#If, also maybe wrap to be consistent with the rest of the docstring..

jonboulle commented on rkt/fetch.go:
```diff
@@ -85,9 +88,38 @@ func runFetch(args []string) (exit int) {
+		ascFile *os.File
+		err     error
+	)
+	if asc != "" && ks != nil {
```
hrm, if asc is supplied but keystore is nil, wonder if we should report this somewhere?

jonboulle commented on rkt/fetch.go:
```diff
@@ -157,26 +197,29 @@ func downloadImage(aciURL string, ascURL string, scheme string, ds *cas.Store, k
-// err will be nil if the ACI downloads successfully and the ACI is verified.
-func download(aciURL string, ascURL string, ds *cas.Store, ks *keystore.Keystore) (*openpgp.Entity, *os.File, error) {
+// fetch opens/downloads and verifies the remote ACI.
+// if ascFile is not nil, it will be used as the signature file and ascURL will be ignored.
```
s/if/If

jonboulle commented on rkt/fetch.go:
```diff
@@ -111,45 +143,53 @@ func fetchImage(img string, ds *cas.Store, ks *keystore.Keystore, discover bool)
+
 	if globalFlags.InsecureSkipVerify {
-		stdout("rkt: warning: signature verification has been disabled")
+		if ks != nil {
```
this is redundant, ks will be nil iff globalFlags.InsecureSkipVerify (maybe we should clean this up a bit)

vcaputo replied:
This was done to keep silent in the stage1.aci ks=nil code path, since stage1 isn't being verified atm.

jonboulle replied:
ok, makes sense for now

jonboulle commented on rkt/fetch.go:
```diff
@@ -85,9 +88,38 @@ func runFetch(args []string) (exit int) {
 	u, err := url.Parse(img)
-	if err == nil && discover && u.Scheme == "" {
+	if err != nil {
+		return "", fmt.Errorf("not a valid URL (%s)", img)
```
not a valid image reference ?

jonboulle commented on rkt/fetch.go:
```diff
@@ -203,43 +246,61 @@ func download(aciURL string, ascURL string, ds *cas.Store, ks *keystore.Keystore
-		defer sigTempFile.Close()
-		defer os.Remove(sigTempFile.Name())
+		if u.Scheme != "file" {
+			stdout("Downloading signature from %v\n", ascURL)
```
can trim the \n

jonboulle commented on rkt/fetch.go:
```diff
@@ -203,43 +246,61 @@ func download(aciURL string, ascURL string, ds *cas.Store, ks *keystore.Keystore
+	if u.Scheme == "file" {
+		aciFile, err = os.Open(u.Path)
+		if err != nil {
+			return nil, nil, fmt.Errorf("error opening aci image: %v", err)
```
damn, we have to get away from "aci image" - just "ACI" (or "ACI file" in this case, please)

jonboulle commented on stage0/run.go:
```diff
@@ -60,19 +61,19 @@ type PrepareConfig struct {
+	SpawnMetadataService bool         // launch metadata service
+	LockFd               int          // lock file descriptor
+	Interactive          bool         // whether the pod is interactive or not
+	Images               []types.Hash // application images (prepare gets them via Apps)
```
I don't get why Run can't also use Apps, and then we could just scrap this field?

vcaputo replied:
It can, but Apps implies so much more than just image ids is being expressed.  Run just needs the image ids because of overlayfs, it doesn't apply anything attached to Apps like prepare does.  I didn't want to create a false expectation in some future reader of this code that manipulating crap in Apps attached to the influences Run, because it doesn't, not right now anyways.

jonboulle commented:
clever!

I like the idea of moving to a structured App to parse into stage0, the previous situation was getting messy.

Per [previous discussion](https://github.com/coreos/rocket/pull/509#issuecomment-84087840) I'm still ambivalent about command lines becoming overly complicated/unintuitive. But if @robszumski is happy from a UX/discoverability perspective then I'm definitely open to persuasion.

vcaputo commented:
@jonboulle all rebased, handles the "--" swallowing in rkt.go, and added the --signature support for rkt fetch (latest commit).

I think I addressed most of the stuff pointed out, lots of rebasing, s/rocket/rkt/g hoorah

merge?  I'd really like this to land so we can move forward on getting the other app-specific flags in place.

jonboulle commented:
@vcaputo there are still a few unaddressed comments from previous reviews, are you planning to fix them in follow up? I am OK with the overall change but would like to see them get addressed

jonboulle commented on common/apps/apps.go:
```diff
@@ -0,0 +1,102 @@
+	apps []App
+}
+
+// Reset creates a new slice for al.apps, needed by tests
```
Methods only used for tests should be declared in test code (it is the same package so should be fine)

jonboulle replied:
no, I lied, it's _not_ the same package, hmm
